octodns/provider/ns1.py

```diff
@@ -275,49 +545,342 @@ def populate(self, zone, target=False, lenient=False):
+        for k, v in expected.items():
+            if have.get(k, '--missing--') != v:
+                return False
+
+        return True
+
+    def _monitor_sync(self, record, value, existing):
+        self.log.debug('_monitor_sync: record=%s, value=%s', record.fqdn,
+                       value)
+        expected = self._monitor_gen(record, value)
+
+        if existing:
+            self.log.debug('_monitor_sync:   existing=%s', existing['id'])
+            monitor_id = existing['id']
+
+            if not self._monitor_is_match(expected, existing):
+                self.log.debug('_monitor_sync:   existing needs update')
+                # Update the monitor to match expected, everything else will be
+                # left alone and assumed correct
+                self._client.monitors_update(monitor_id, **expected)
+
+            feed_id = self._client.feeds_for_monitors.get(monitor_id)
+            if feed_id is None:
+                self.log.warn('_monitor_sync: %s (%s) missing feed, creating',
+                              existing['name'], monitor_id)
+                feed_id = self._feed_create(existing)
+        else:
+            self.log.debug('_monitor_sync:   needs create')
+            # We don't have an existing monitor create it (and related bits)
+            monitor_id, feed_id = self._monitor_create(expected)
+
+        return monitor_id, feed_id
+
+    def _monitors_gc(self, record, active_monitor_ids=None):
+        self.log.debug('_monitors_gc: record=%s, active_monitor_ids=%s',
+                       record.fqdn, active_monitor_ids)
+
+        if active_monitor_ids is None:
+            active_monitor_ids = set()
+
+        for monitor in self._monitors_for(record).values():
+            monitor_id = monitor['id']
+            if monitor_id in active_monitor_ids:
+                continue
+
+            self.log.debug('_monitors_gc:   deleting %s', monitor_id)
+
+            feed_id = self._client.feeds_for_monitors.get(monitor_id)
+            if feed_id:
+                self._client.datafeed_delete(self._client.datasource_id,
+                                             feed_id)
+
+            self._client.monitors_delete(monitor_id)
+
+            notify_list_id = monitor['notify_list']
+            self._client.notifylists_delete(notify_list_id)
+
+    def _params_for_dynamic_A(self, record):
+        pools = record.dynamic.pools
+
+        # Convert rules to regions
+        regions = {}
+        for i, rule in enumerate(record.dynamic.rules):
+            pool_name = rule.data['pool']
+
+            notes = {
+                'rule-order': i,
+            }
+
+            fallback = pools[pool_name].data.get('fallback', None)
+            if fallback:
+                notes['fallback'] = fallback
+
+            country = set()
+            georegion = set()
+            us_state = set()
+
+            for geo in rule.data.get('geos', []):
+                n = len(geo)
+                if n == 8:
```

Comment:
Comment examples inline of the strings would be nice here to help remind the reader what we're parsing.

Reply:
Done.